python/ffcc/io.py

```diff
@@ -172,26 +180,39 @@ def _list_subfolders(path):
     return subfolders
 
   folders = [os.path.join(path, f) for f in _list_subfolders(path)]
-  data = []
+  # include the main directory
+  folders.append(path)
+  image_files = []
   for folder in folders:
-    try:
-      illuminant, weight, frames = read_from_directory(folder)
-      for i, f in enumerate(frames):
-        rgb, extended_feature = f
-        data.append({
-            'name': '{}_{}'.format(os.path.basename(folder), i),
-            'rgb': rgb,
-            'extended_feature': extended_feature,
-            'illuminant': illuminant,
-            'weight': weight,
-        })
-    except:  # pylint: disable=bare-except
-      tf.logging.error('Unable to load %s', folder)
-
-  return data
-
-
-def write_dataset_to_tfrecord(dataset, output_dir):
-  """Encodes dataset into TFRecords and write into the destination folder."""
-  for d in dataset:
-    encode_to_tfrecord(d, output_dir)
+    if folder:
+      image_files = [
+        os.path.join(folder, file) for file in os.listdir(folder) if
+        file.endswith('.png')]
+
+  # check if cvfolds.txt file exists in the root path
+  if tf.io.gfile.exists(os.path.join(path, 'cvfolds.txt')):
+    cvfolds = np.loadtxt(os.path.join(path, 'cvfolds.txt'), dtype=np.int)
+  else:
+    # create a 3-fold cross-validation partition and store it in the root path
+    number_of_files = len(image_files)
+    indices = np.linspace(0, number_of_files - 1, number_of_files).astype(int)
+    cvfolds = np.zeros(number_of_files).astype(int)
+    random.shuffle(indices)
+    for fold in range(3):
+      cvfolds[indices[np.floor((fold * number_of_files) / 3).astype(int):
+                      np.floor(((fold + 1) * number_of_files) / 3).astype(int)
+              ]] = int(fold + 1)
+    cvfolds_file = open(os.path.join(path, 'cvfolds.txt'), 'w')
+    np.savetxt(cvfolds_file, cvfolds, fmt='%d')
+    cvfolds_file.seek(cvfolds_file.tell() - 1, os.SEEK_SET)
+    cvfolds_file.truncate()
+    cvfolds_file.close()
+
+  training_files, eval_files = get_training_eval_sets(
+    image_files, cvfolds, test_fold)
+
+  training_input, training_label = build_dataset_dict(training_files, shuffle)
+
+  eval_input, eval_label = build_dataset_dict(eval_files, shuffle)
+
+  return training_input, training_label, eval_input, eval_label
```

Comment:
Note the 'no newline at end of file' warning symbol in github; you should add newlines at the end of each file. I think you should be able to configure your editor to enforce this for you.